### PR TITLE
IoT.js Module Generator

### PR DESCRIPTION
**Pull request by fbmrk**
The purpose of this tool to create a native iotjs module from a C / C ++ API by generating the binding layer. It's still in progress because the C ++ part has not been completed yet.

For more information and specification please read `docs/devs/IoT.js-Module-Generator.md`

IoT.js-DCO-1.0-Signed-off-by: Marko Fabo mfabo@inf.u-szeged.hu

### CODE REVIEW

**Comment by yichoi** [comment]
Looks promising! It would be better to provide a novel way to avoid memory leaks at this generator.

**Comment by fbmrk** [comment]
What do you mean exactly? Are there problems with jerry values or the direct `malloc` calls in the binding layer?

**Comment by yichoi** [comment]
It's general problem that Native layer and JS layer cannot trace each others memory life time exactly therefore memory could be leaked. I am just throwing my idea, please don't be distracted. :)

**Comment by daeyeon** [comment]
Interesting idea. Comparing the other binding generating idea like a conventional IDL-based one, users might feel familiar because it's C/C++ language-based.

We'd like to know what the header for binding looks like for a _stateful_ module such as `Buffer`. Can you add the example also?

**Comment by fbmrk** [comment]
I have updated the PR and made some refactoring. I have also add an example binding layer to `docs/devs/IoT.js-Module-Generator.md`.

The C++ part is still in progress.

**Comment by fbmrk** [comment]
I have updated the PR. The tool has reached a usable state and it gives some help for developers to avoid writing a lots of code. It doesn't solve the whole problem of create a binding between C/C++ and JavaScript languages, but in simple cases it can fully do the task of the developer and works well. The tool can't handle perfect when there are complex user-defined types, pointers to structures, multiple indirections etc. in the native library.
For more information please check the documentation: `docs/devs/IoT.js-Module-Generator.md`

File: tools/iotjs-generate-module.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python
+
+# Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
```
Comment (rtakacs) [inline tools/iotjs-generate-module.py]: The year should be updated.

Comment (fbmrk) [reply tools/iotjs-generate-module.py]: Done, thank you

**Comment by daeyeon** [comment]
1. Can you move the example (`/test_c`, `/text_cpp`, and `test.py`) into `test/{new folder}`?
2. In `docs/devs/IoT.js-Module-Generator.md`, it seems better to mention the current limitation you mentioned above.

**Comment by fbmrk** [comment]
@daeyeon Thank you for your comments. I've updated the PR.

**Pull request merged**